smart_open/smart_open_lib.py

```diff
@@ -372,82 +394,105 @@ class ParseUri(object):
-                # access key id: [A-Z0-9]{20}
-                # secret access key: [A-Za-z0-9/+=]{40}
-                acc, self.bucket_id = self.bucket_id
-                self.access_id, self.access_secret = acc.split(':')
-                self.bucket_id, self.key_id = self.bucket_id.split('/', 1)
-            elif len(self.bucket_id) == 3 and len(self.bucket_id[0].split(':')) == 2:
-                # or URI in extended format: s3://key:secret@server[:port]@bucket/object
-                acc,  server, self.bucket_id = self.bucket_id
-                self.access_id, self.access_secret = acc.split(':')
-                self.bucket_id, self.key_id = self.bucket_id.split('/', 1)
-                server = server.split(':')
-                self.ordinary_calling_format = True
-                self.host = server[0]
-                if len(server) == 2:
-                    self.port = int(server[1])
-            else:
-                # more than 2 '@' means invalid uri
-                # Bucket names must be at least 3 and no more than 63 characters long.
-                # Bucket names must be a series of one or more labels.
-                # Adjacent labels are separated by a single period (.).
-                # Bucket names can contain lowercase letters, numbers, and hyphens.
-                # Each label must start and end with a lowercase letter or a number.
-                raise RuntimeError("invalid S3 URI: %s" % uri)
-        elif self.scheme == 'file':
-            self.uri_path = parsed_uri.netloc + parsed_uri.path
-
-            # '~/tmp' may be expanded to '/Users/username/tmp'
-            self.uri_path = os.path.expanduser(self.uri_path)
-
-            if not self.uri_path:
-                raise RuntimeError("invalid file URI: %s" % uri)
-        elif self.scheme.startswith('http'):
-            self.uri_path = uri
-        else:
-            raise NotImplementedError("unknown URI scheme %r in %r" % (self.scheme, uri))
+    if os.name == 'nt':
+        # urlsplit doesn't work on Windows -- it parses the drive as the scheme...
+        if '://' not in uri_as_string:
+            # no protocol given => assume a local file
+            uri_as_string = 'file://' + uri_as_string
+    parsed_uri = urlsplit(uri_as_string, allow_fragments=False)
+
+    if parsed_uri.scheme == "hdfs":
+        return _parse_uri_hdfs(parsed_uri)
+    elif parsed_uri.scheme == "webhdfs":
+        return _parse_uri_webhdfs(parsed_uri)
+    elif parsed_uri.scheme in ("s3", "s3n", "s3u"):
+        return _parse_uri_s3x(parsed_uri)
+    elif parsed_uri.scheme in ('file', '', None):
+        return _parse_uri_file(parsed_uri)
+    elif parsed_uri.scheme.startswith('http'):
+        return Uri(scheme=parsed_uri.scheme, uri_path=uri_as_string)
+    else:
+        raise NotImplementedError(
+            "unknown URI scheme %r in %r" % (parsed_uri.scheme, uri_as_string)
+        )
+
+
+def _parse_uri_hdfs(parsed_uri):
+    assert parsed_uri.scheme == 'hdfs'
```

Comment:
Why asserts needed (you check exact same conditions before in `if`/`elif`), here and everywhere? Sanity-check only?

Reply:
Yes, it's a sanity check only.